### DISCUSSION TITLE
PlugData v0.5.3

### DISCUSSION THREAD

**Discussion by timothyschoen** [plugdata-team/plugdata#144] [General]
Big release with many changes, I've tried to list them all but there's probably even more. Performance should be much better. Large parts of the GUI object code have been rewritten and redesigned, many of the bugs that v0.5.2 suffered from are fixed. Because of the large rewrites, you may still encounter some bugs, please report them to me!

On Linux, I recommend installing it through the [package repository](https://software.opensuse.org//download.html?project=home%3Aplugdata&package=plugdata) (thanks @KottV), or the [AUR package](https://aur.archlinux.org/packages/plugdata-git) on Arch (thanks @aik2mlj).

Thanks for reporting bugs and suggesting features:  @vewilya, @TjShredder, @60-hz, @zkf, @TheGuyWhoo, @sevdh, @alfonso73,  @plantappreciator, @ludnny, @jaffasplaffa, @offthesky, @JoshuaACNewman

Sorry if I forgot anyone, this release has been in development for a while!

Thanks to @JoshuaACNewman and @ludnny for also sponsoring the development of PlugData, I appreciate it!

- Updated cyclone to 6.0-1
- Updated ELSE to 1.0-rc2 (sfont~ object coming next release... sorry!)
- Updated to JUCE 7
- Fixed glitch when dragging close to resizer
- Inlets/outlets hide by default unless hovered
- Inlets/outlets become greyed out when locked
- Fixed Reaper windows focus issue when using other dialog in Reaper
- Fixed jumpiness for grid lines
- Fixed mouse position problems in large patches
- Can set some dynamic parameters (some don't work yet, I'm working on it!):
- Fixed crash when renaming some objects
- Console correctly groups lists on one line
- Fixed some console bugs
- Fixed several object resizing bugs
- Fixed freeze in reaper
- Fixed copy issue
- Fixed potential startup crash
- Can now send automation (i'll explain how soon on the wiki)
- Better look for DAW audio settings sliders and automation panel sliders
- Added about panel in settings
- Can now set a custom font
- Improved comment object size
- Better text wrapping for all objects
- Limit for numberbox height
- Fix cases where connection hitbox blocks inlet/outlet
- Reconnect handles for connections
- Now can move objects with arrow keys
- Fixed auto fontsize
- Fixed crash when loading preset
- Fixed freeze when opening empty file
- Fixed issues with keyboard object
- Mian canvas now has properties (by right-clicking)
- Fixed dollsym in label bug
- Fixed Atom send/receive problems
- Array improvements
    - Dratically improved performance
    - Fixed programatically resizing
    - Fixed freeze when loading large arrays
    - Now has title
    - Fixed drag glitches
    - Now has range property
- Improved console performance
- Better default colour for canvas object
- Fixed many gui details
- Fixed problems with invalid objects
- Fixed wrong default label position
- Fixed label position bug when objects have negative coordinates
- Fixed atom label, fixed atom height issues
- Optimised startup time
- Message resize while typing
- Improved lasso performance
- Ability to modify the colour scheme
- Command-line args
    - -path "path"     -- add to file search path
    - -open "file>"    -- open file(s) on startup
    - -send "msg..."   -- send a message at startup, after patches are loaded
- Fixed locale issue for countries that use comma instead of period for numbers
- Subpatch issue
- Multiple grid snap directions
- Now can use ELSE and cyclone objects with name prefixes
- Help patch shortcut
- Set textobjectwidth functie
    - Sla width op in chars
    - Bereken zelf width!
- Fixed some of the connectionstate bugs
- Object size more similar to Pd
- Better object UI
- Added Message box mouse-down animation
- Added listbox dragging
- Hide subpatchers and abstractions in presentation mode
- Cmd-s shortcut problem
- Fixed slowness on Windows
- Fixed some document browser problems
- Fixed deken crash
- Fixed crash when deleting inlets/outlets from subpatcher
- Fixed some of the issues with automation
- Infinite numbox dragging
- Better support for pd's drawing functions
- Fixed no MIDI output when called from pd's thread
- Correct latency compensation
- Fixed crash when pinch-zooming
- Fixed attach-to-mouse problem when creating object from menu

Known Issues:
- Broken objects: [text define], [netreceive], [store], [rec], [oscilloscope], possibly more.
- LV2 doesn't work in Ardour, this is Ardour's fault. Use the VST3 instead.
- C-coded externals are broken on windows

<hr /><em>This discussion was created from the release <a href='https://github.com/timothyschoen/PlugData/releases/tag/v0.5.3'>PlugData v0.5.3</a>.</em>

**Comment by KottV** [comment]
Great!

**Comment by alfonso73** [comment]
yeah!

**Comment by FlachyJoe** [comment]
Great but I'm still waiting for [netreceive] to test with [FCPDWorkbench](https://github.com/FlachyJoe/FCPDWorkbench) :wink:

**Reply by timothyschoen** [reply]
I'm going to fix all of the vanilla and ELSE objects that are currently broken (that I know of) for the next release!

**Reply by timothyschoen** [reply]
What about netreceive isn't working for you? I've tested it, and I can successfully send messages between Max and PlugData:

https://user-images.githubusercontent.com/44585538/178125930-2af81c9a-5ff8-4f95-84b7-de03648f39fe.mov

**Reply by timothyschoen** [reply]
Aha, so if you use netreceive in the DAW, it only works if audio is running. I'll see what I can do about that.

**Reply by timothyschoen** [reply]
Fixed!

**Reply by timothyschoen** [reply]
Nope! I have a fix for this but currently, it's too unstable. It might make it into a later release. For now, objects like netreceive will only work as long as the DAW is sending audio blocks to PlugData. This also applies to other objects where timing is important, like [metro] and many others: they will only work if audio is running.

**Comment by alfonso73** [comment]
As are you workin' on smart patching could you add a feature to select all cables only so would be possibile to convert all of them in segmented form and back to curved? thanks!

**Reply by timothyschoen** [reply]
Selected objects are ignored when changing connection style, so ctrl-a should probably do the job?

**Comment by alfonso73** [comment]
<img width="298" alt="NOT OK" src="https://user-images.githubusercontent.com/3230434/177140143-3f8803c1-e800-452b-9a62-f6c0fec4573b.png">
<img width="292" alt="OK" src="https://user-images.githubusercontent.com/3230434/177140151-858e50c5-bb43-4c8c-9f97-a98438c0315e.png">

**Comment by alfonso73** [comment]
@timothyschoen as you can see selection is kinda strange....if i select some part the blue icon for segmented patch wires is available...with another selection is greyed out....kinda strange

**Reply by timothyschoen** [reply]
Oh yeah that is kinda strange, the problem is that I wasn't sure how to handle the button's state when you have connections of both types selected. A workaround is to use the shortcut ctrl-shift-g to turn them all into segmented connections and go from there. But I agree, this is bad, I'll find a better way to handle this.

**Comment by alfonso73** [comment]
Seen that you added bandlimited osc to ELSE! That's a great improvement! Do you think you and Alexandre can add band limited sine, cosine and impulse oscillators to? Would be great for FM/PM/PD synthesis and various west-coast style stuff.

**Reply by timothyschoen** [reply]
We're planning to add impulse oscillators, I don't think there are plans for band-limited sines yet. bl.sine~ should do the trick, but it will be a lot more processor intensive. You can also use PlugData's new built-in oversampling:


<img width="243" alt="Screenshot 2022-07-22 at 17 36 33" src="https://user-images.githubusercontent.com/44585538/180474193-892c74c9-7b5a-469a-8fe0-138cb2124740.png">

**Reply by alfonso73** [reply]
Uh that's for efficient and bandlimited FM/PD/PM synthesis. Probably in the meantime a good patching workaround would be using bandlimited triangle osc with a frequency controlled filter to turn a triangle in quasi-sine

**Comment by TjShredder** [comment]
Isn't a sine by definition bandlimited? Anyway, I see this project connected to Pd, and its focus should be to make Pd work in a DAW context. To include specific objects, for example bandlimited oscillators, is up to the main branch...
I am more interested if PlugData would have a CLAP version, it might even streamline developement:
https://github.com/free-audio/clap

**Reply by alfonso73** [reply]
@TjShredder if the sine Is modulated via PM, FM, PD synthesis is not a sine anymore and need to be bandlimited

**Reply by timothyschoen** [reply]
EDIT: misread CLAP as CREB, oops

I see the point of having a band-limited sine wave, it's just that I have no idea how to do that. Bandlimiting a saw or square is simple because the aliasing is mostly caused by the discontinuity in the waveform. We can replace that discontinuity with a band-limited discontinuity and voila! However, the aliasing of a modulated sine wave is not caused by discontinuities, so it's much more difficult to get rid of that without oversampling.

From this you can also conclude that the other waveshapes are also not completely alias-free if you apply FM/PM, but getting rid of the discontinuities still gets us 90% of the way there. If you want to have perfect FM/PM/PD, oversampling could still be the best way to go. I believe that most FM synths either oversample, or adjust the FM parameters to make sure there is no aliasing.

**Reply by timothyschoen** [reply]
About a CLAP version, PlugData is based on JUCE, so I don't maintain plugin formats manually. It's up to the JUCE devs to implement CLAP. I think they will do it eventually, but it may still be a while.

**Reply by JoshuaACNewman** [reply]
@alfonso73 THANK you! I've been furiously trying to understand!

**Reply by alfonso73** [reply]
Thanks @timothyschoen. I'll do some research on this interesting topic ASAP.

**Comment by sevdh** [comment]
Very nice!

Congratulations!

On 29 Jun 2022, at 00:36, Timothy Schoen ***@***.***> wrote:

﻿
Big release with many changes, I've tried to list them all but there's probably even more. Performance should be much better. Large parts of the GUI object code have been rewritten and redesigned, many of the bugs that v0.5.2 suffered from are fixed. Because of the large rewrites, you may still encounter some bugs, please report them to me!

On Linux, I recommend installing it through the package repository (thanks @KottV), or the AUR package on Arch (thanks @aik2mlj).

Thanks for reporting bugs and suggesting features: @vewilya, @TjShredder, @60-hz, @zkf, @TheGuyWhoo, @sevdh, @alfonso73, @plantappreciator, @ludnny, @jaffasplaffa, @offthesky, @JoshuaACNewman

Sorry if I forgot anyone, this release has been in development for a while!

Thanks to @JoshuaACNewman and @ludnny for also sponsoring the development of PlugData, I appreciate it!

Updated cyclone to 6.0-1
Updated ELSE to 1.0-rc2 (sfont~ object coming next release... sorry!)
Updated to JUCE 7
Fixed glitch when dragging close to resizer
Inlets/outlets hide by default unless hovered
Inlets/outlets become greyed out when locked
Fixed Reaper windows focus issue when using other dialog in Reaper
Fixed jumpiness for grid lines
Fixed mouse position problems in large patches
Can set some dynamic parameters (some don't work yet, I'm working on it!):
Fixed crash when renaming some objects
Console correctly groups lists on one line
Fixed some console bugs
Fixed several object resizing bugs
Fixed freeze in reaper
Fixed copy issue
Fixed potential startup crash
Can now send automation (i'll explain how soon on the wiki)
Better look for DAW audio settings sliders and automation panel sliders
Added about panel in settings
Can now set a custom font
Improved comment object size
Better text wrapping for all objects
Limit for numberbox height
Fix cases where connection hitbox blocks inlet/outlet
Reconnect handles for connections
Now can move objects with arrow keys
Fixed auto fontsize
Fixed crash when loading preset
Fixed freeze when opening empty file
Fixed issues with keyboard object
Mian canvas now has properties (by right-clicking)
Fixed dollsym in label bug
Fixed Atom send/receive problems
Array improvements
Dratically improved performance
Fixed programatically resizing
Fixed freeze when loading large arrays
Now has title
Fixed drag glitches
Now has range property
Improved console performance
Better default colour for canvas object
Fixed many gui details
Fixed problems with invalid objects
Fixed wrong default label position
Fixed label position bug when objects have negative coordinates
Fixed atom label, fixed atom height issues
Optimised startup time
Message resize while typing
Improved lasso performance
Ability to modify the colour scheme
Command-line args
-path "path" -- add to file search path
-open "file>" -- open file(s) on startup
-send "msg..." -- send a message at startup, after patches are loaded
Fixed locale issue for countries that use comma instead of period for numbers
Subpatch issue
Multiple grid snap directions
Now can use ELSE and cyclone objects with name prefixes
Help patch shortcut
Set textobjectwidth functie
Sla width op in chars
Bereken zelf width!
Fixed some of the connectionstate bugs
Object size more similar to Pd
Better object UI
Added Message box mouse-down animation
Added listbox dragging
Hide subpatchers and abstractions in presentation mode
Cmd-s shortcut problem
Fixed slowness on Windows
Fixed some document browser problems
Fixed deken crash
Fixed crash when deleting inlets/outlets from subpatcher
Fixed some of the issues with automation
Infinite numbox dragging
Better support for pd's drawing functions
Fixed no MIDI output when called from pd's thread
Correct latency compensation
Fixed crash when pinch-zooming
Fixed attach-to-mouse problem when creating object from menu
Known Issues:

Broken objects: [text define], [netreceive], [store], [rec], [oscilloscope], possibly more.
LV2 doesn't work in Ardour, this is Ardour's fault. Use the VST3 instead.
C-coded externals are broken on windows
This discussion was created from the release PlugData v0.5.3.
